### PR TITLE
Discovery

### PR DESCRIPTION
**Pull request by mensfeld**
@agwozdowski  @pavlo-vavruk have fun :)

### CODE REVIEW

File: lib/karafka/connection/broker.rb [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+    # @note By using OpenStruct as a base we can use JSON data to map directly to object attributes
+    class Broker < OpenStruct
+      # @param json_data [String] string with JSON data describing broker
+      def initialize(json_data)
+        super JSON.parse(json_data)
+      end
+
+      # @return [String] host with port for this broker
+      # @example
+      #   broker #=> '172.16.0.1:9092'
```
Comment (pavlo-vavruk) [inline lib/karafka/connection/broker.rb]: `url #=> '172.16.0.1:9092'`

File: spec/lib/karafka/connection/broker_manager_spec.rb [hunk 2]
```diff
@@ -0,0 +1,85 @@
+require 'spec_helper'
+
+RSpec.describe Karafka::Connection::BrokerManager do
+  subject { described_class.new }
```
Comment (pavlo-vavruk) [inline spec/lib/karafka/connection/broker_manager_spec.rb]: redundant

**Comment by agwozdowski** [comment]
:+1:

File: README.md [hunk 3]
```diff
@@ -134,6 +134,10 @@ There are several env settings you can use:
 | KARAFKA_ENV       | development     | In what mode this application should boot (production/development/test/etc)   |
 | KARAFKA_BOOT_FILE | app_root/app.rb | Path to a file that contains Karafka app configuration and booting procedures |
 
+### Kafka brokers auto-discovery
+
+Karafka supports Kafka brokers auto-discovery during both startup and runtime. It means that **zookeeper_hosts** option allows Karafka to get all the details it needs about Kafka brokers, first during boot and after each failure. If something happens to a connection on which we werelistening (or if we cannot connect to a given broker), Karafka will refresh list of available brokers. This allows it to be aware of changes that happen in the infrastructure (adding and removing nodes) and allows it to be up and running as long as Zookeeper is able to provide it all the required informations.
+
```
Comment (pavlo-vavruk) [inline README.md]: werelistening => were listening
informations => information

File: lib/karafka/config.rb [hunk 4]
```diff
@@ -19,6 +18,7 @@ class << self
     # option wait_timeout [Integer] seconds that we will wait on a single topic for messages
     # option worker_timeout [Integer] how many seconds should we proceed stuff at Sidekiq
     # option zookeeper_hosts [Array] zookeeper hosts with ports where zookeeper servers are run
+    # option kafka_hosts [Array] - optional - kafka hosts with ports (autodiscovered if missing)
     SETTINGS = %i(
       logger
       kafka_hosts
```
Comment (pavlo-vavruk) [inline lib/karafka/config.rb]: Move here to the end of the settings list as well

File: spec/lib/karafka/connection/broker_manager_spec.rb [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+  describe '#zk' do
+    let(:zk_instance) { double }
+
+    it 'expect to create ZK instance out of joined hosts' do
+      expect(ZK)
+        .to receive(:new)
+        .with(::Karafka::App.config.zookeeper_hosts.join(','))
+        .and_return(zk_instance)
+
+      subject.send(:zk)
```
Comment (pavlo-vavruk) [inline spec/lib/karafka/connection/broker_manager_spec.rb]: redundant

File: lib/karafka/connection/broker.rb [hunk 6]
```diff
@@ -0,0 +1,20 @@
+module Karafka
+  module Connection
+    # Object representing a single Kafka broker (node) information
+    # It is being built based on Zooekeeper incoming json data
```
Comment (pavlo-vavruk) [inline lib/karafka/connection/broker.rb]: Zooekeeper => Zookeeper

File: lib/karafka/config.rb [hunk 7]
```diff
@@ -35,6 +35,18 @@ class << self
       attr_accessor attr_name
     end
 
+    # @return [Array<String>] all Kafka hosts (with porsts)
```
Comment (pavlo-vavruk) [inline lib/karafka/config.rb]: portsts => ports

**Pull request merged**